libexec/python/shell.py

```diff
@@ -99,63 +187,66 @@ def parse_image_uri(image,
     if default_tag is None:
         default_tag = TAG
 
-    # Be absolutely sure there are not comments
+    # if user gave custom registry / namespace, make them the default
+    if CUSTOM_NAMESPACE is not None:
+        default_namespace = CUSTOM_NAMESPACE
+
+    if CUSTOM_REGISTRY is not None:
+        default_registry = CUSTOM_REGISTRY
+
+    # candidate regex for matching, in order of preference
+    uri_regexes = [ _reduced_uri_no_ns_re,
+                    _default_uri_re ]
+
+    # Be absolutely sure there are no comments
     image = image.split('#')[0]
 
-    # Get rid of any uri, and split the tag
-    image = image.replace(uri, '')
-
-    # Does the uri have a digest or Github tag (version)?
-    image = image.split('@')
-    version = None
-    if len(image) == 2:
-        version = image[1]
-
-    image = image[0]
-    image = image.split(':')
-
-    # If there are three parts, we have port and tag
-    if len(image) == 3:
-        repo_tag = image[2]
-        image = "%s:%s" % (image[0], image[1])
-
-    # If there are two parts, we have port or tag
-    elif len(image) == 2:
-        # If there isn't a slash in second part, we have a tag
-        if image[1].find("/") == -1:
-            repo_tag = image[1]
-            image = image[0]
-        # Otherwise we have a port and we merge the path
-        else:
-            image = "%s:%s" % (image[0], image[1])
-            repo_tag = default_tag
-    else:
-        image = image[0]
-        repo_tag = default_tag
+    if not uri:
+        uri = get_image_uri(image, quiet=True)
 
-    # Now look for registry, namespace, repo
-    image = image.split('/')
+    # docker images require slightly different rules
+    if uri == "docker://":
+        uri_regexes = [ _docker_uri_re ]
 
-    if len(image) > 2:
-        registry = image[0]
-        namespace = "/".join(image[1:-1])
-        repo_name = image[-1]
+    image = remove_image_uri(image, uri);
 
-    elif len(image) == 2:
-        registry = default_registry
-        namespace = image[0]
-        repo_name = image[1]
+    for r in uri_regexes:
+        match = r.match(image)
+        if match:
+            break
 
-    else:
+    if not match:
+        bot.error('Could not parse image "%s"! Exiting.' % image)
+        sys.exit(1)
+
+    registry = match.group('registry')
+    namespace = match.group('namespace')
+    repo_name = match.group('repo')
+    repo_tag = match.group('tag')
+    version = match.group('version')
+
+    if namespace:
+        # strip trailing /
+        namespace = namespace.rstrip('/')
+
+    # repo_name is required and enforced by the re (in theory)
+    # if not provided, re should not match
+    assert(repo_name)
```

Comment:
This is a good test! Can we catch this `AssertionError` and give a meaningful message to the user? I think as of now it would just throw the error and exit.

Reply:
This is a "impossible" condition, in theory. The regex should enforce that, if it matched at all, the `repo_name` should always be present, and if not it's a bug. The user should never have to deal with this, and for developers, the stack trace is probably the most useful error message.
I'm not sure if I understood what you meant correctly though, how do you normally handle these cases in singularity?

Reply:
Oh, what I mean is that, if the condition fails (which it probably wouldn't) the user would see a stack trace. A developer is probably ok with this, but we've had a few issues in the past of people complaining about seeing the stack trace instead of a nice clean message. so something like:

```
try:
    assert(repo_name)
except AssertionError:
    bot.error('A repository name is required.')
    sys.exit(1)
```

Reply:
Well, you might as well do
```python
if not repo_name:
    bot.error('A repository name is required')
    sys.exit(1)
```
in that case, but please hear me out on what I say below.

First of all, I'm not in charge here and will ultimately do what I'm told, but this is my opinion: an end user who knows nothing about the internals of singularity will have no use for that error message, and if anything, it will mislead him/her to think they didn't specify the repository, when maybe they did but there was a bug in singularity.
If a user encounters this, what you want from them is to report it here so it can be fixed. Maybe dumping a trace in their face is not the best way to ask them to help, but maybe something like capturing the stack trace into a file and showing a message like:
```
Internal error at file.py:123. Stack trace dumped at /tmp/singularity.bugreport.
Please report a bug at https://github.com/singularityware/singularity/issues and attach the file above.
```
would serve the purpose better? And I even argue this should be caught at the top level of the program, not locally in the function itself, so it can catch all the asserts generically.

Reply:
I can give you examples of the issues that I'm referring to, if it helps:

 - https://github.com/singularityware/singularity/issues/880
 - https://github.com/singularityware/singularity/issues/751#issuecomment-308243851

(I didn't mean to target that user, I just remembered his username and having reported a few).

I think either way the users come to post an issue, and a meaningful message (and you can do what you think is best) is preferable to a dump of python, which for someone that doesn't know python, might as well be Chinese :) 

We also want to exit with status 1 so the calling script gets the message too :)